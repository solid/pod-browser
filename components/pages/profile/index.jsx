/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, { useContext } from "react";
import { SessionContext } from "@inrupt/solid-ui-react";
import { useRouter } from "next/router";
import Profile from "../../profile";
import EditableProfile from "../../profile/editableProfile";
import useFullProfile from "../../../src/hooks/useFullProfile";

export default function ProfileShow() {
  const router = useRouter();
  const { session, sessionRequestInProgress } = useContext(SessionContext);

  const decodedIri = router.query.webId
    ? decodeURIComponent(router.query.webId)
    : null;

  const agentProfile = useFullProfile(decodedIri ?? session.info.webId);

  if (sessionRequestInProgress || !agentProfile) {
    return null;
  }
  const editableProfile = agentProfile?.editableProfileDatasets[0];

  if (editableProfile) {
    return (
      <EditableProfile
        profile={agentProfile}
        profileDataset={editableProfile}
        webId={agentProfile.webId}
      />
    );
  }
  return <Profile profile={agentProfile} webId={decodedIri} />;
}
