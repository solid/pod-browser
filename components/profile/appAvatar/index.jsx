/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, { useState } from "react";
import { foaf } from "rdf-namespaces";
import { Avatar, Box, createStyles } from "@material-ui/core";
import { makeStyles } from "@material-ui/styles";
import { useBem } from "@solid/lit-prism-patterns";
import { Icons } from "@inrupt/prism-react-components";
import {
  Text,
  ThingProvider,
  DatasetProvider,
  Image,
} from "@inrupt/solid-ui-react";

import styles from "./styles";
import {
  mockApp,
  mockAppDataset,
  LOGO_PREDICATE,
} from "../../../__testUtils/mockApp";

const useStyles = makeStyles((theme) => createStyles(styles(theme)));

export const TESTCAFE_ID_NAME_TITLE = "app-name-title";
export const TESTCAFE_ID_NAME_FIELD = "app-name-field";
export const TESTCAFE_ID_WEBID_FIELD = "app-webid-field";
export const TESTCAFE_ID_TOS_FIELD = "app-tos-field";
export const TESTCAFE_ID_POLICY_FIELD = "app-policy-field";

export function setupErrorComponent(bem) {
  return () => (
    <Avatar className={bem("appAvatar")} alt="Contact avatar">
      <Icons className={bem("appAvatar")} name="project-diagram" />
    </Avatar>
  );
}

// temporarily using mock data for apps for dev purposes until we have audit list
const app = mockApp();
const dataset = mockAppDataset();

export default function AppAvatar() {
  const [error, setError] = useState(null);

  const classes = useStyles();
  const bem = useBem(classes);
  // TODO replace with toast error or something?
  if (error) {
    return error.toString();
  }

  const errorComponent = setupErrorComponent(bem);

  return (
    <DatasetProvider solidDataset={dataset}>
      <ThingProvider thing={app} onError={setError}>
        <Box alignItems="center" display="flex">
          <Box>
            <Avatar className={classes.appAvatar} alt="Contact avatar">
              <Image
                property={LOGO_PREDICATE}
                width={120}
                errorComponent={errorComponent}
              />
            </Avatar>
          </Box>

          <Box p={2}>
            <h3 data-testid={TESTCAFE_ID_NAME_TITLE}>
              <Text property={foaf.name} />
            </h3>
          </Box>
        </Box>
      </ThingProvider>
    </DatasetProvider>
  );
}
