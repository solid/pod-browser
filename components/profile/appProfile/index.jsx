/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, { useState } from "react";
import { foaf, vcard } from "rdf-namespaces";
import clsx from "clsx";
import { getThing, getUrl } from "@inrupt/solid-client";
import {
  Avatar,
  Box,
  Paper,
  InputLabel,
  createStyles,
} from "@material-ui/core";
import { makeStyles } from "@material-ui/styles";
import { useBem } from "@solid/lit-prism-patterns";
import {
  Container,
  Icons,
  Table as PrismTable,
} from "@inrupt/prism-react-components";
import {
  Text,
  ThingProvider,
  DatasetProvider,
  Value,
  Image,
  Table,
  TableColumn,
} from "@inrupt/solid-ui-react";

import { vcardExtras } from "../../../src/addressBook";

import styles from "./styles";
import {
  mockApp,
  mockAppDataset,
  TOS_PREDICATE,
  POLICY_PREDICATE,
  CONTACTS_PREDICATE,
  LOGO_PREDICATE,
} from "../../../__testUtils/mockApp";
import ValueBody from "./valueCellBody";
import TypeBody from "./titleCellBody";

const useStyles = makeStyles((theme) => createStyles(styles(theme)));

export const TESTCAFE_ID_NAME_TITLE = "app-name-title";
export const TESTCAFE_ID_NAME_FIELD = "app-name-field";
export const TESTCAFE_ID_WEBID_FIELD = "app-webid-field";
export const TESTCAFE_ID_TOS_FIELD = "app-tos-field";
export const TESTCAFE_ID_POLICY_FIELD = "app-policy-field";

export function setupErrorComponent(bem) {
  return () => (
    <Avatar className={bem("appAvatar")} alt="Contact avatar">
      <Icons className={bem("appAvatar")} name="project-diagram" />
    </Avatar>
  );
}

// temporarily using mock data for apps for dev purposes until we have audit list
const app = mockApp();
const dataset = mockAppDataset();

export default function AppProfile() {
  const [error, setError] = useState(null);
  const tableClass = PrismTable.useTableClass("table", "inherits");

  const classes = useStyles();
  const bem = useBem(classes);
  // TODO replace with toast error or something?
  if (error) {
    return error.toString();
  }

  const errorComponent = setupErrorComponent(bem);
  // FIXME: not sure what the contacts would look like in a real app, but possibly a dataset with things? Maybe a profile? For now it is an array of strings:
  const contactsUrl = getUrl(app, CONTACTS_PREDICATE);
  const contacts = getThing(dataset, contactsUrl);

  return (
    <Container>
      <Paper style={{ marginTop: "1em" }}>
        <Box p={2}>
          <DatasetProvider solidDataset={dataset}>
            <ThingProvider thing={app} onError={setError}>
              <Box alignItems="center" display="flex">
                <Box>
                  <Avatar className={classes.appAvatar} alt="Contact avatar">
                    <Image
                      property={LOGO_PREDICATE}
                      width={120}
                      errorComponent={errorComponent}
                    />
                  </Avatar>
                </Box>

                <Box p={2}>
                  <h3 data-testid={TESTCAFE_ID_NAME_TITLE}>
                    <Text property={foaf.name} />
                  </h3>
                </Box>
              </Box>

              <hr />

              <Box mt={2}>
                <Box>
                  <InputLabel>App WebID</InputLabel>
                  <Value
                    property={vcardExtras("WebId")}
                    dataType="url"
                    autosave
                    data-testid={TESTCAFE_ID_WEBID_FIELD}
                  />
                </Box>
              </Box>
              <Box mt={2}>
                <Box>
                  <InputLabel>Terms of Service</InputLabel>
                  <Value
                    property={TOS_PREDICATE}
                    dataType="url"
                    autosave
                    data-testid={TESTCAFE_ID_TOS_FIELD}
                  />
                </Box>
              </Box>
              <Box mt={2}>
                <Box>
                  <InputLabel>Privacy Policy</InputLabel>
                  <Value
                    property={POLICY_PREDICATE}
                    dataType="url"
                    autosave
                    data-testid={TESTCAFE_ID_POLICY_FIELD}
                  />
                </Box>
              </Box>
            </ThingProvider>
            <ThingProvider thing={contacts} onError={setError}>
              <Box mt={2}>
                <InputLabel>Contact Support</InputLabel>
                <Table
                  className={clsx(tableClass, bem("table"))}
                  things={[
                    { dataset, thing: contacts },
                    { dataset, thing: contacts },
                    { dataset, thing: contacts },
                  ]}
                >
                  <TableColumn
                    property={vcard.url}
                    body={({ row }) => <TypeBody id={row.id} />}
                    header=""
                  />
                  <TableColumn
                    dataType="url"
                    property={vcard.url}
                    header=""
                    body={({ row }) => <ValueBody id={row.id} />}
                  />
                </Table>
              </Box>
            </ThingProvider>
          </DatasetProvider>
        </Box>
      </Paper>
    </Container>
  );
}
