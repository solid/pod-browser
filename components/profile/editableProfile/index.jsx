/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import T from "prop-types";
import { Box, Paper } from "@material-ui/core";
import { Container } from "@inrupt/prism-react-components";
import { CombinedDataProvider } from "@inrupt/solid-ui-react";
import { getThing } from "@inrupt/solid-client";
import { schema } from "rdf-namespaces";
import PersonProfile from "../personProfile";
import PersonAvatar from "../personAvatar";
import AppProfile from "../appProfile";
import { profilePropTypes } from "../../../constants/propTypes";

export const TESTCAFE_ID_NAME_TITLE = "profile-name-title";
export const TESTCAFE_ID_NAME_FIELD = "profile-name-field";
export const TESTCAFE_ID_ROLE_FIELD = "profile-role-field";
export const TESTCAFE_ID_ORG_FIELD = "profile-org-field";

export default function EditableProfile({ profileDataset, profile, webId }) {
  if (!profileDataset) {
    return <span>No profile document found this this WebID</span>;
  }
  if (profile.types.includes(schema.SoftwareApplication)) {
    return <AppProfile />;
  }

  return (
    <CombinedDataProvider solidDataset={profileDataset} thingUrl={webId}>
      <Container>
        <Paper style={{ marginTop: "1em" }}>
          <Box p={2}>
            <PersonAvatar
              profileDataset={profileDataset}
              profile={profile}
              editing
            />
            <hr />
            <PersonProfile
              profileDataset={profileDataset}
              profile={profile}
              editing
            />
          </Box>
        </Paper>
      </Container>
    </CombinedDataProvider>
  );
}

EditableProfile.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  profileDataset: T.object,
  profile: profilePropTypes,
  webId: T.string.isRequired,
};

EditableProfile.defaultProps = {
  profileDataset: null,
  profile: null,
};
