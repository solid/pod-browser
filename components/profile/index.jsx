/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, { useState } from "react";
import T from "prop-types";
import { Box, Paper } from "@material-ui/core";
import { Alert } from "@material-ui/lab";
import { Container } from "@inrupt/prism-react-components";
import { CombinedDataProvider } from "@inrupt/solid-ui-react";
import { schema } from "rdf-namespaces";
import { isHTTPError } from "../../src/error";
import PersonProfile from "./personProfile";
import PersonAvatar from "./personAvatar";
import AppProfile from "./appProfile";

export const TESTCAFE_ID_NAME_TITLE = "profile-name-title";
export const TESTCAFE_ID_NAME_FIELD = "profile-name-field";
export const TESTCAFE_ID_ROLE_FIELD = "profile-role-field";
export const TESTCAFE_ID_ORG_FIELD = "profile-org-field";

export default function Profile(props) {
  const { profileIri, editing, type } = props;
  const [error, setError] = useState(null);

  // TODO replace with toast error or something?
  if (error) {
    if (isHTTPError(error, 404)) {
      return (
        <Alert severity="error">
          {`Cannot fetch avatar for this WebID: ${profileIri}`}
        </Alert>
      );
    }
    return error.toString();
  }

  if (type === schema.SoftwareApplication) {
    return <AppProfile />;
  }

  return (
    <CombinedDataProvider
      datasetUrl={profileIri}
      thingUrl={profileIri}
      onError={setError}
    >
      <Container>
        <Paper style={{ marginTop: "1em" }}>
          <Box p={2}>
            <PersonAvatar profileIri={profileIri} />
            <hr />
            <PersonProfile profileIri={profileIri} editing={editing} />
          </Box>
        </Paper>
      </Container>
    </CombinedDataProvider>
  );
}

Profile.propTypes = {
  profileIri: T.string.isRequired,
  editing: T.bool,
  type: T.string.isRequired,
};

Profile.defaultProps = {
  editing: false,
};
