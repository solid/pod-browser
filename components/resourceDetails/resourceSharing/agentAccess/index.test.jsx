/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";

import { waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { createAccessMap } from "../../../../src/solidClientHelpers/permissions";
import AgentAccess from "./index";
import { renderWithTheme } from "../../../../__testUtils/withTheme";
import * as profileFns from "../../../../src/solidClientHelpers/profile";
import { mockProfileAlice } from "../../../../__testUtils/mockPersonResource";
import { PUBLIC_AGENT_PREDICATE } from "../../../../src/models/contact/public";
import { AUTHENTICATED_AGENT_PREDICATE } from "../../../../src/models/contact/authenticated";
import useAgentProfile from "../../../../src/hooks/useAgentProfile";

jest.mock("../../../../src/hooks/useAgentProfile");
const mockedUseAgentProfile = useAgentProfile;

const webId = "https://example.com/profile/card#me";

describe("AgentAccess", () => {
  describe("with profile", () => {
    beforeEach(() => {
      mockedUseAgentProfile.mockReturnValue({ data: mockProfileAlice() });
    });
    const permission = {
      acl: createAccessMap(true, true, false, false),
      webId,
      alias: "Editors",
      type: "agent",
    };

    it("renders", () => {
      const { asFragment } = renderWithTheme(
        <AgentAccess permission={permission} />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
  describe("for public agent", () => {
    const permission = {
      acl: createAccessMap(true, true, false, false),
      webId: PUBLIC_AGENT_PREDICATE,
      alias: "Editors",
      type: "public",
    };
    it("renders correctly", () => {
      const { asFragment } = renderWithTheme(
        <AgentAccess permission={permission} />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
  describe("for authenticated agent", () => {
    const permission = {
      acl: createAccessMap(true, true, false, false),
      webId: AUTHENTICATED_AGENT_PREDICATE,
      alias: "Editors",
      type: "authenticated",
    };
    it("renders correctly", () => {
      const { asFragment } = renderWithTheme(
        <AgentAccess permission={permission} />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
  describe("without profile", () => {
    const mutateProfile = jest.fn();
    it("renders skeleton placeholders when profile is not available", () => {
      beforeEach(() => {
        mockedUseAgentProfile.mockReturnValue({
          data: undefined,
          error: new Error("error"),
          mutate: mutateProfile,
        });
      });
      const { asFragment } = renderWithTheme(
        <AgentAccess
          permission={{
            acl: createAccessMap(true, true, false, false),
            webId,
            alias: "Editors",
            type: "agent",
          }}
        />
      );

      expect(asFragment()).toMatchSnapshot();
    });
    it("returns null when no access", () => {
      const { asFragment } = renderWithTheme(
        <AgentAccess
          permission={{
            acl: null,
            webId,
            alias: "Editors",
            type: "agent",
          }}
        />
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it("renders a skeleton while loading profile", async () => {
      mockedUseAgentProfile.mockReturnValue({
        data: undefined,
        error: undefined,
      });
      const { asFragment } = renderWithTheme(
        <AgentAccess
          permission={{
            acl: createAccessMap(true, true, false, false),
            webId,
            alias: "Editors",
            type: "agent",
          }}
          mutatePermissions={mutatePermissions}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("renders an error message with a 'try again' button if it's unable to load profile", () => {
      const { asFragment, findByTestId } = renderWithTheme(
        <AgentAccess
          permission={{
            acl: createAccessMap(true, true, false, false),
            webId,
            alias: "Editors",
            type: "agent",
          }}
        />
      );
      expect(findByTestId("try-again-button")).toBeTruthy();
      expect(asFragment()).toMatchSnapshot();
    });

    it("renders a spinner after clicking 'try again' button", async () => {
      const { findByTestId } = renderWithTheme(
        <AgentAccess
          permission={{
            acl: createAccessMap(true, true, false, false),
            webId,
            alias: "Editors",
            type: "agent",
          }}
        />
      );
      const button = await findByTestId("try-again-button");
      userEvent.click(button);

      expect(findByTestId("try-again-spinner")).toBeTruthy();
    });

    it("tries to fetch the profile again when clicking 'try again' button", async () => {
      const fetchProfileSpy = jest.spyOn(profileFns, "fetchProfile");
      const { findByTestId } = renderWithTheme(
        <AgentAccess
          permission={{
            acl: createAccessMap(true, true, false, false),
            webId,
            alias: "Editors",
            type: "agent",
          }}
        />
      );
      const button = await findByTestId("try-again-button");
      userEvent.click(button);

      await waitFor(() =>
        expect(fetchProfileSpy).toHaveBeenCalledWith(webId, expect.anything())
      );
    });

    it("removes the spinner when fetching succeeds", async () => {
      const fetchProfileSpy = jest.spyOn(profileFns, "fetchProfile");

      const { findByTestId, queryByTestId } = renderWithTheme(
        <AgentAccess
          permission={{
            acl: createAccessMap(true, true, false, false),
            webId,
            alias: "Editors",
            type: "agent",
          }}
        />
      );
      const button = await findByTestId("try-again-button");
      userEvent.click(button);
      mockedUseAgentProfile.mockReturnValue({
        data: mockProfileAlice(),
      });
      await waitFor(() =>
        expect(fetchProfileSpy).toHaveBeenCalledWith(webId, expect.anything())
      );
      fetchProfileSpy.mockResolvedValueOnce(mockProfileAlice());
      await waitFor(() => expect(mutateProfile).toHaveBeenCalled());
      await waitFor(() =>
        expect(queryByTestId("try-again-spinner")).toBeFalsy()
      );
    });

    it("removes the spinner when fetching errors", async () => {
      const fetchProfileSpy = jest.spyOn(profileFns, "fetchProfile");
      const { getByTestId, queryByTestId } = renderWithTheme(
        <AgentAccess
          permission={{
            acl: createAccessMap(true, true, false, false),
            webId,
            alias: "Editors",
            type: "agent",
          }}
        />
      );
      await waitFor(() => {
        const button = getByTestId("try-again-button");
        userEvent.click(button);
      });

      await waitFor(() =>
        expect(fetchProfileSpy).toHaveBeenCalledWith(webId, expect.anything())
      );
      await waitFor(() =>
        expect(queryByTestId("try-again-spinner")).toBeFalsy()
      );
    });
  });
});
