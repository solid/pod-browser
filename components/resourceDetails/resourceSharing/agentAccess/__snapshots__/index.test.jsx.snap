// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AgentAccess for authenticated agent renders correctly 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-nameAndAvatarContainer"
  >
    <div
      class="PodBrowser-root PodBrowser-circle PodBrowser-avatar PodBrowser-colorDefault"
    >
      <span
        class="PodBrowser-icon-user-lock PodBrowser-avatar__icon PodBrowser-avatar__icon--authenticated"
      />
    </div>
    <p
      class="PodBrowser-root PodBrowser-detailText PodBrowser-body1 PodBrowser-detailText"
      data-testid="agent-web-id"
    >
      Anyone signed in
    </p>
    <button
      class="PodBrowser-button"
      data-testid="menu-button"
      type="button"
    >
      <i
        class="PodBrowser-icon-more"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`AgentAccess for public agent renders correctly 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-nameAndAvatarContainer"
  >
    <div
      class="PodBrowser-root PodBrowser-circle PodBrowser-avatar PodBrowser-colorDefault"
    >
      <span
        class="PodBrowser-icon-globe PodBrowser-avatar__icon PodBrowser-avatar__icon--public"
      />
    </div>
    <p
      class="PodBrowser-root PodBrowser-detailText PodBrowser-body1 PodBrowser-detailText"
      data-testid="agent-web-id"
    >
      Anyone
    </p>
    <button
      class="PodBrowser-button"
      data-testid="menu-button"
      type="button"
    >
      <i
        class="PodBrowser-icon-more"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`AgentAccess with profile renders 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-nameAndAvatarContainer"
  >
    <div
      class="PodBrowser-root PodBrowser-circle PodBrowser-avatar"
    >
      <img
        alt="https://example.com/profile/card#me"
        class="PodBrowser-img"
        src="http://alice.example.com/alice.jpg"
      />
    </div>
    <p
      class="PodBrowser-root PodBrowser-detailText PodBrowser-body1 PodBrowser-detailText"
      data-testid="agent-web-id"
    >
      Alice
    </p>
    <button
      class="PodBrowser-button"
      data-testid="menu-button"
      type="button"
    >
      <i
        class="PodBrowser-icon-more"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`AgentAccess without profile renders a skeleton while loading profile 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-loadingStateContainer"
  >
    <span
      class="PodBrowser-root PodBrowser-circle PodBrowser-pulse"
      style="width: 40px; height: 40px;"
    />
    <div>
      <span
        class="PodBrowser-root PodBrowser-text PodBrowser-pulse"
        style="width: 100px;"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`AgentAccess without profile renders an error message with a 'try again' button if it's unable to load profile 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-alert-container"
  >
    <div
      class="PodBrowser-root PodBrowser-root PodBrowser-alertBox PodBrowser-standardWarning PodBrowser-elevation0"
      role="alert"
    >
      <div
        class="PodBrowser-icon PodBrowser-icon"
      >
        <svg
          aria-hidden="true"
          class="PodBrowser-root PodBrowser-fontSizeInherit"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 5.99L19.53 19H4.47L12 5.99M12 2L1 21h22L12 2zm1 14h-2v2h2v-2zm0-6h-2v4h2v-4z"
          />
        </svg>
      </div>
      <div
        class="PodBrowser-message PodBrowser-alertMessage"
      >
        Unable to load this profile
      </div>
      <div
        class="PodBrowser-action PodBrowser-action"
      >
        <button
          class="PodBrowser-root PodBrowser-root PodBrowser-text PodBrowser-bold-button PodBrowser-colorInherit PodBrowser-textSizeSmall PodBrowser-sizeSmall"
          data-testid="try-again-button"
          tabindex="0"
          type="button"
        >
          <span
            class="PodBrowser-label"
          >
            Try again
          </span>
          <span
            class="PodBrowser-root"
          />
        </button>
      </div>
    </div>
    <div
      class="PodBrowser-separator"
    />
    <div
      class="PodBrowser-nameAndAvatarContainer"
    >
      <div
        class="PodBrowser-root PodBrowser-circle PodBrowser-avatar PodBrowser-colorDefault"
      >
        <span
          class="PodBrowser-icon-user PodBrowser-avatar__icon"
        />
      </div>
      <p
        class="PodBrowser-root PodBrowser-detailText PodBrowser-body1 PodBrowser-detailText"
        data-testid="agent-web-id"
      >
        https://example.com/profile/card#me
      </p>
      <button
        class="PodBrowser-button"
        data-testid="menu-button"
        type="button"
      >
        <i
          class="PodBrowser-icon-more"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AgentAccess without profile renders skeleton placeholders when profile is not available 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-loadingStateContainer"
  >
    <span
      class="PodBrowser-root PodBrowser-circle PodBrowser-pulse"
      style="width: 40px; height: 40px;"
    />
    <div>
      <span
        class="PodBrowser-root PodBrowser-text PodBrowser-pulse"
        style="width: 100px;"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`AgentAccess without profile returns null when no access 1`] = `<DocumentFragment />`;
